src/marks/tip.js

```diff
@@ -332,18 +334,73 @@ function getSources({channels}) {
   return sources;
 }
 
-function formatPair(c1, c2, i) {
+function formatTitle(i, index, {title}) {
+  const format = this.format?.title;
+  return format === null ? [] : (format ?? formatDefault)(title.value[i], i);
+}
+
+function* formatChannels(i, index, channels, scales, values) {
+  for (const key in channels) {
+    if (key === "x1" && "x2" in channels) continue;
+    if (key === "y1" && "y2" in channels) continue;
+    const channel = channels[key];
+    if (key === "x2" && "x1" in channels) {
+      const format = this.format?.x; // TODO x1, x2?
+      if (format === null) continue;
+      yield {
+        label: formatPairLabel(scales, channels, "x"),
+        value: formatPair(format ?? formatDefault, channels.x1, channel, i)
+      };
+    } else if (key === "y2" && "y1" in channels) {
+      const format = this.format?.y; // TODO y1, y2?
+      if (format === null) continue;
+      yield {
+        label: formatPairLabel(scales, channels, "y"),
+        value: formatPair(format ?? formatDefault, channels.y1, channel, i)
+      };
+    } else {
+      const format = this.format?.[key];
+      if (format === null) continue;
```

Comment:
Possible future improvement: we could prune the result of getSources based on this.format; then we wouldn’t need to skip here. Also, we could materialize the formatDefault as needed.